include/libopencm3/stm32/h7/rcc.h

```diff
@@ -181,23 +328,228 @@ LGPL License Terms @ref lgpl_license
+#define RCC_D2CCIP2R_USARTSEL_PLL3Q         2
+#define RCC_D2CCIP2R_USARTSEL_HSI           3
+#define RCC_D2CCIP2R_USARTSEL_CSI           4
+#define RCC_D2CCIP2R_USARTSEL_LSE           5
+#define RCC_D2CCIP2R_USARTSEL_MASK          7
+/**@}*/
+
+
+#define HSI_BASE_FREQUENCY              64000000UL
+
+/** Enumerations for clocks in the clock tree to allow user to get the current configuration of the
+ *  clocks from the RCC module. These clock sources will each be tracked through the settings.
+ */
+enum rcc_clock_source {
+  RCC_CPUCLK,
+  RCC_SYSCLK,
+  RCC_PERCLK,
+  RCC_SYSTICKCLK,
+  RCC_HCLK3,
+  RCC_AHBCLK,           /* AHB1,2,4 all share base HCLK. */
+  RCC_APB1CLK,          /* APB1 and PCLK1 are aliases. */
+  RCC_PCLK1,
+  RCC_APB2CLK,
+  RCC_PCLK2,
+  RCC_APB3CLK,
+  RCC_PCLK3,
+  RCC_APB4CLK,
+  RCC_PCLK4,
+  RCC_FDCAN1CLK,        /* FDCAN controllers share a kernel clock */
+  RCC_FDCAN2CLK,
+  RCC_SPI1CLK,          /* SPI 1/2/3 share a kernel clock. */
+  RCC_SPI2CLK,
+  RCC_SPI3CLK,
+  RCC_SPI4CLK,          /* SPI 4/5 share kernel clock. */
+  RCC_SPI5CLK,
+  RCC_SPI6CLK,          /* SPI6 has its own kernel clock. */
+  RCC_USART1CLK,        /* USART 1/6 share kernel clock. */
+  RCC_USART6CLK,
+  RCC_USART2CLK,        /* USART 2/3/4/5/6/8 share kernel clock. */
+  RCC_USART3CLK,
+  RCC_USART4CLK,
+  RCC_USART5CLK,
+  RCC_USART7CLK,
+  RCC_USART8CLK,
+};
 
 enum rcc_osc {
   RCC_PLL,
   RCC_HSE,
   RCC_HSI,
-  RCC_LSE,
-  RCC_LSI
+  RCC_LSE
+};
+
+enum rcc_sysclk_mux {
+  RCC_SYSCLK_PLL,
+  RCC_SYSCLK_HSE,
+  RCC_SYSCLK_HSI,
+};
+
+enum rcc_pll_mux {
+  RCC_PLL_HSI = RCC_PLLCKSELR_PLLSRC_HSI,
+  RCC_PLL_HSE = RCC_PLLCKSELR_PLLSRC_HSE
+};
+
+enum rcc_d1cpre {
+  RCC_D1CPRE_BYP = RCC_D1CFGR_D1CPRE_BYP << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV2 = RCC_D1CFGR_D1CPRE_DIV2 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV4 = RCC_D1CFGR_D1CPRE_DIV4 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV8 = RCC_D1CFGR_D1CPRE_DIV8 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV16 = RCC_D1CFGR_D1CPRE_DIV16 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV64 = RCC_D1CFGR_D1CPRE_DIV64 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV128 = RCC_D1CFGR_D1CPRE_DIV128 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV256 = RCC_D1CFGR_D1CPRE_DIV256 << RCC_D1CFGR_D1CPRE_SHIFT,
+  RCC_D1CPRE_DIV512 = RCC_D1CFGR_D1CPRE_DIV512 << RCC_D1CFGR_D1CPRE_SHIFT,
+};
+
+enum rcc_d1ppre {
+  RCC_D1PPRE_BYP = RCC_D1CFGR_D1PPRE_BYP << RCC_D1CFGR_D1PPRE_SHIFT,
+  RCC_D1PPRE_DIV2 = RCC_D1CFGR_D1PPRE_DIV2 << RCC_D1CFGR_D1PPRE_SHIFT,
+  RCC_D1PPRE_DIV4 = RCC_D1CFGR_D1PPRE_DIV4 << RCC_D1CFGR_D1PPRE_SHIFT,
+  RCC_D1PPRE_DIV8 = RCC_D1CFGR_D1PPRE_DIV8 << RCC_D1CFGR_D1PPRE_SHIFT,
+  RCC_D1PPRE_DIV16 = RCC_D1CFGR_D1PPRE_DIV16 << RCC_D1CFGR_D1PPRE_SHIFT,
+};
+
+enum rcc_d1hpre {
+  RCC_D1HPRE_BYP = RCC_D1CFGR_D1HPRE_BYP,
+  RCC_D1HPRE_DIV2 = RCC_D1CFGR_D1HPRE_DIV2,
+  RCC_D1HPRE_DIV4 = RCC_D1CFGR_D1HPRE_DIV4,
+  RCC_D1HPRE_DIV8 = RCC_D1CFGR_D1HPRE_DIV8,
+  RCC_D1HPRE_DIV16 = RCC_D1CFGR_D1HPRE_DIV16,
```

Comment:
I gotta say, all these just feels super spammy.  We've never put these wrappers in elsewhere, I don't really see why we need them here.  Everyone else just gets a uint32 for params.  IF, (and it's a big if) you want to move apis all over to be enums, I would say the enums should just _be_ the definitions, and would still be named accordingly, instead of this duplicate-but-not pasting.

Reply:
I did this specifically for the configuration parameters that go into the pll config struct only to constrain the selections, and allow the compiler/static analyzers/code indexers/etc. to let the user know when they've used the wrong value in the wrong place (not that I've ever done it before or anything :-)). The macros are used elsewhere, so I found having both definitions useful.  

The struct could be changed to be all _uint32_t_ values, but I thought this a pretty narrow area where adding a bit more explicit typing would be beneficial.

As you've noted below, I didn't go "all out" on this with everything, just the config struct. I can see the argument to all/none, so I will go ahead and eliminate the enums.

Reply:
Went ahead and removed these...added some helper macros to make creating a config struct not so painful.